### PR TITLE
Biolucida image view

### PR DESCRIPTION
**Pull request by hsorby**
# Description

Update Biolucida viewer to open images in Biolucida app, add permalink button for Biolucida images.

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Staged here: https://mapcore-demo.org/staging/sparc-app/
Manually tested.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have utilized components from the Design System Library where applicable
- [ ] I have added unit tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Review by alan-wu — COMMENTED** [review]
Looks good in general, I found some bugs regarding to the scaffoldvuer and have patched them up in your branch. I am not able to get the share link resuming to where it should be with the biolucida image viewer on the test site you posted.

I like the changes on file details, I think we can go even further into making it a component itself instead of shared css and mixin but what you have done is good.

**Comment by alan-wu** [comment]
I have tested with few different images, it looks like it sets the position but it does not set the slice of image it was on.
If I get a new share link on a link with a query location set already, it will create am extra location instead of replacing it. (&location=-165,402,1,45&location=157,-77,1,45 for example)

**Comment by alan-wu** [comment]
Looks like the wrong slice number is sent to us. If I set the slice number manually on the address bar, the viewer will go to the correct slice so the implementation is fine.

File: components/BiolucidaViewer/BiolucidaViewer.vue [hunk 1]
```diff
@@ -33,38 +31,61 @@ export default {
...
-      // console.log('received message.')
-      // console.log(event.origin)
-      // console.log(event.data)
+      if (event.origin === 'https://sparc.biolucida.net') {
+        const message = event.data
+        if (message === 'setting x,y,z,f failed') {
+          this.$message(failMessage('Unable to set image location.'))
+        } else if (!message.startsWith('setting x,y,z,f ')) {
+          this.$copyText(
+            `${process.env.ROOT_URL}${this.$route.fullPath}&location=${message}`
```
Comment (alan-wu) [inline components/BiolucidaViewer/BiolucidaViewer.vue:78]: Should replace the location query instead.

File: components/BiolucidaViewer/BiolucidaViewer.vue [hunk 2]
```diff
@@ -1,20 +1,16 @@
...
-          </bf-button>
-        </a>
-      </p>
-      <iframe ref="biolucida" :src="data.share_link" />
+      <el-row class="mb-2 justify-space-between" type="flex">
+        <bf-button class="ml-8 btn-copy-permalink-solid" @click="launchViewer">
+          View in 3D
+        </bf-button>
+        <bf-button class="" @click="queryView">
+          <svg-icon name="icon-permalink" height="24" width="24" />
```
Comment (Tehsurfer) [inline components/BiolucidaViewer/BiolucidaViewer.vue:9]: If we want to match the other parts of the site, I would move the permalink to here and have it be icon only

image

Can see an example of permalinks elsewhere here:
https://sparc.science/news-and-events/events/RifYKQymUFDJymzyyehm5

image

**Review by Tehsurfer — CHANGES_REQUESTED** [review]
When I run this branch locally the permalink button does not copy the link for send any messages which is strange. It works on the demo site you have given. 

I also put a few other questions I had as comments

File: components/ImagesGallery/ImagesGallery.vue [hunk 3]
```diff
@@ -574,7 +573,8 @@ export default {
             this.$set(item, 'thumbnail', this.defaultImg)
           }
 
-          return Promise.reject('Maximum iterations reached.')
+          console.log('Maximum iterations reached aborting.')
```
[the remark this answers is not in the document]
Comment (hsorby) [reply components/ImagesGallery/ImagesGallery.vue:576]: Yes. Using this as a gentle way to remind about search result errors causing problems.

File: components/BiolucidaViewer/BiolucidaViewer.vue (hunk 2 above)
Comment (hsorby) [reply components/BiolucidaViewer/BiolucidaViewer.vue:9]: I have matched the button styling but have not moved it.  I'll let UX decide if it should be moved.

File: assets/_viewer.scss [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+  border-bottom: 1px solid #dbdfe6;
+  flex-direction: column;
+  font-size: 0.875em;
+  display: flex;
+  padding: 1rem 0.625rem;
+  @media (min-width: 48em) {
+    flex-direction: row;
+  }
+}
+.file-detail__column {
```
[the remark this answers is not in the document]
Comment (hsorby) [reply assets/_viewer.scss:13]: Not sure how this would help. we want the full column width available.

**Comment by hsorby** [comment]
I have now addressed review comments.

File: components/BiolucidaViewer/BiolucidaViewer.vue [hunk 5]
```diff
@@ -33,44 +37,73 @@ export default {
...
+      this.$refs.biolucida.contentWindow.postMessage(
+        'getImgPos',
+        'https://sparc.biolucida.net'
+      )
+    },
+    biolucidaLoaded() {
+      if (this.data.location) {
+        this.$refs.biolucida.contentWindow.postMessage(
+          this.data.location,
+          'https://sparc.biolucida.net'
```
Comment (Tehsurfer) [inline components/BiolucidaViewer/BiolucidaViewer.vue:73]: Has the api ever changed? 

Could maybe be an envVar if you could be bothered

**Review by Tehsurfer — CHANGES_REQUESTED** [review]
Main thing is need finish merging with the new sparc-app/master changes

**Review by Tehsurfer — APPROVED** [review]
looks good to me

**Pull request merged**